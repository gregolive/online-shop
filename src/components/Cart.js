import '../styles/Cart.css';
import '@fortawesome/fontawesome-free/js/all';

const Cart = ({ open, items, handleClick }) => {
  const containerClass = (open) ? 'CartContainer Open' : 'CartContainer Close';

  const cartProducts = (
    (items && items.length > 0) ?
    <div className='CartProducts'>
      {items.map((product) =>
        <div>
          {product.name}
        </div>
      )}
    </div>
    : null
  );

  return (
    <div className={containerClass}>
      <div className='Cart'>
        <header className='CartHeader'>
          <div className='CartTitle'>
            <button type='button' className='Button TransparentButton' onClick={handleClick}>
              <i className='fa-solid fa-xmark'></i>
            </button>
            <h4>Shopping cart</h4>
          </div>
          <p className='CartCount'>0 items</p>
        </header>

        {cartProducts}

        
      </div>
    </div>
  );
};

export default Cart;
