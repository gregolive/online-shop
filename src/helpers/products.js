import uniqid from 'uniqid';

import pearlImg from '../img/Acaia-Pearl-S-black.jpg';
import tampImg from '../img/forcetamp585.jpg';
import basketImg from '../img/ims18g-ridgeless.jpg';
import knockboxImg from '../img/maple-knock-set.jpg';
import burrImg from '../img/ssp-multi-purpose.jpg';
import matImg from '../img/tamping-mat-corner.jpg';

class Product {
  constructor (name, vendor, price, stock, img) {
    this.id = uniqid();
    this.name = name;
    this.vendor = vendor;
    this.price = this.formatPrice(price);
    this.stock = stock;
    this.img = img;
  }

  inStock() { return (this.stock > 0) ? true : false };

  formatPrice = (value) => {
    const dollars = Math.trunc(value / 100);
    let cents = value % 100;
    if (cents < 10) { cents = '0' + cents; }
    return `$${dollars}.${cents}`
  };
};

const products = [
  new Product('Pearl Model S Coffee Scale - Black', 'Acaia', 19500, 5, pearlImg),
  new Product('The Force Tamp 58.5mm', 'Espresso Parts', 19900, 2, tampImg),
  new Product('Precision Portafilter Basket - Double 18g - Ridgeless', 'IMS', 3058, 4, basketImg),
  new Product('Maplewood Knockbox Set', 'IMS', 6396, 10, knockboxImg),
  new Product('Multi-Purpose Grinding Red Speed Coated 64mm Flat Burrs', 'SSP', 18500, 1, burrImg),
  new Product('Tamping Mat with Corner Edge', 'Barista Basics', 2800, 0, matImg),
];

export default products;

const findProduct = (productName) => products.filter((p) => p.name === productName)[0];

export { findProduct };
