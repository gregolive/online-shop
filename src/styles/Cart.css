.CartContainer {
  background: rgba(0, 0, 0, 0.5);
  height: 100vh;
  width: 0;
  position: fixed;
  z-index: 1;
  top: 0;
  right: 0;
  overflow-x: hidden;
}

.CartContainer.Close {
  transition: 0.2s;
}

.CartContainer.Open {
  width: 100%;
}

.Cart {
  color: var(--text);
  background: white;
  min-height: 100vh;
  width: 0;
  display: flex;
  flex-direction: column;
  float: right;
  transition: 0.4s;
}

.CartContainer.Open .Cart {
  width: min(80vw, 30rem);
}

.CartHeader {
  background: var(--primary);
  padding: 0 6vw;
  height: 5rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.CartTitle {
  display: flex;
  align-items: center;
}

.CartTitle > Button {
  font-size: 1.5rem;
  padding: 0;
  margin-right: 2rem;
  align-self: start;
}

h4 {
  font-size: 1.8rem;
  font-weight: 700;
  margin: 0;
}

.CartCount {
  color: var(--highlight);
  font-size: 1.2rem;
}

.CartProducts {
  padding: 2rem 6vw;
  display: grid;
}

.CartRow {
  display: grid;
  grid-template: 5rem / auto 1fr;
  row-gap: 0.5rem;
}

.CartRow:not(:last-child) {
  padding-bottom: 1rem;
  border-bottom: 1px solid var(--text);
  margin-bottom: 1rem;
}

.CartRow > img {
  height: 5rem;
}

.ItemInfo {
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5rem;
}

.ItemInfo > small {
  font-size: 0.9rem;
}

h5 {
  font-size: 1rem;
  font-weight: 700;
  margin: 0;
}

.ItemInfo > strong {
  font-size: 1rem;
  font-weight: 600;
}

.ItemQuantity {
  grid-column: 2 / -1;
  display: grid;
  grid-template: 1.5rem / repeat(3, 1.5rem) 5rem;
  justify-items: center;
  align-items: center;
}

.ItemQuantity > Button {
  padding: 0.25rem;
}
