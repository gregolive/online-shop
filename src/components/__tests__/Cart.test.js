import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Cart from '../Cart';
import products from '../../helpers/Product';
import CartItem from '../../helpers/CartItem';

const testProduct = products[0];
const testItem = new CartItem(testProduct, 2)
const testItemCount = testItem.quantity;

test('renders close button only when no item is added', () => {
  render(<MemoryRouter><Cart /></MemoryRouter>);
  expect(screen.getAllByRole('button').length).toEqual(1);
});

test('renders 5 total buttons when one type of item is added', () => {
  render(<MemoryRouter><Cart items={[testItem]} itemCount={testItemCount}/></MemoryRouter>);
  expect(screen.getAllByRole('button').length).toEqual(5);
});

test('renders heading', () => {
  render(<MemoryRouter><Cart /></MemoryRouter>);
  expect(screen.getByRole('heading').textContent).toMatch(/Shopping cart/);
});

test('renders number of items', () => {
  render(<MemoryRouter><Cart items={[testItem]} itemCount={testItemCount}/></MemoryRouter>);
  expect(screen.getByText(/2 items/)).toBeInTheDocument();
});

test('renders product image', () => {
  render(<MemoryRouter><Cart items={[testItem]} itemCount={testItemCount}/></MemoryRouter>);
  expect(screen.getByAltText(`${testProduct.vendor} - ${testProduct.name}`)).toBeInTheDocument();
});

test('renders product vendor', () => {
  render(<MemoryRouter><Cart items={[testItem]} itemCount={testItemCount}/></MemoryRouter>);
  expect(screen.getByText(testProduct.vendor)).toBeInTheDocument();
});

test('renders product name', () => {
  render(<MemoryRouter><Cart items={[testItem]} itemCount={testItemCount}/></MemoryRouter>);
  expect(screen.getByText(testProduct.name)).toBeInTheDocument();
});

test('renders proceed to checkout button with proper text', () => {
  render(<MemoryRouter><Cart items={[testItem]} itemCount={testItemCount}/></MemoryRouter>);
  expect(screen.getAllByRole('button')[4].textContent).toMatch(/Proceed to Checkout/)
});

it('renders consistently', () => {
  const { container } = render(<Cart />);
  expect(container).toMatchSnapshot();
});
