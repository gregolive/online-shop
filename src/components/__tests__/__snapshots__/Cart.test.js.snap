// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders consistently 1`] = `
<div>
  <div
    class="CartContainer"
  >
    <div
      class="Cart"
    >
      <header
        class="CartHeader"
      >
        <div
          class="CartTitle"
        >
          <button
            class="Button TransparentButton"
            type="button"
          >
            <i
              class="fa-solid fa-xmark"
            />
          </button>
          <h4>
            Shopping cart
          </h4>
        </div>
        <p
          class="CartCount"
        >
           
          items
        </p>
      </header>
      <div
        class="CartProducts"
      />
    </div>
  </div>
</div>
`;
